### PR TITLE
FEAT(Day10): add enclosed_tiles

### PR DESCRIPTION
**Pull request by dariannax3**
The implementation of the above function forced to add another important function color_neighbors

During implementation few bugs were discovered and fixed in function:
- generate_all_surroudning_tiles_starting_from_left_top_clock_wise problem was with loop tiles bordering the vector boundaries
- find_next_neighbors as above
- is_in_maze - vecotr sizes were swaped

Co-authored-by: Daria Kumanek <dariakumanek@maladie.local>

### CODE REVIEW

File: day_10/task_2/BUILD [hunk 1]
```diff
@@ -3,14 +3,17 @@ cc_binary(
     srcs = ["main.cpp"],
     deps = [
         ":pipe_maze",
-        "//path_helper"
-        # "//debug_features:debug_features"],
+        "//path_helper",
+        "//debug_features:debug_features"
```
Comment (Rosein) [inline day_10/task_2/BUILD]:
```suggestion
```
debug features should be removed before the merge

File: day_10/task_2/BUILD [hunk 2]
```diff
@@ -3,14 +3,17 @@ cc_binary(
...
 
 cc_library(
     name = "pipe_maze",
     hdrs = ["pipe_maze.hpp"],
     srcs = ["pipe_maze.cpp"],
-    visibility = ["//visibility:public"]
+    visibility = ["//visibility:public"],
+    deps = [
+        "//debug_features:debug_features"
+    ]
```
Comment (Rosein) [inline day_10/task_2/BUILD]: debug features should be removed before the merge
```suggestion
```

File: day_10/task_2/pipe_maze.cpp [hunk 3]
```diff
@@ -211,16 +210,22 @@ void AttributedMaze::set_tile_at(const MazePoint& point, const Tile& tile)
...
+    return point.first >= 0 and point.first < static_cast<int>(maze_.size()) and point.second >= 0 and point.second < static_cast<int>(maze_[0].size());
 }
 
 MazePoint operator+(const MazePoint& lhs, const MazePoint& rhs)
 {
     return MazePoint{lhs.first + rhs.first, lhs.second + rhs.second};
 }
 
-std::vector<MazePoint> generate_all_surroudning_tiles_starting_from_left_top_clock_wise(const MazePoint& mid_loop_point)
+void AttributedMaze::removeInvalidMazePoints(std::vector<MazePoint>& tiles) {
```
Comment (Rosein) [inline day_10/task_2/pipe_maze.cpp]:
```suggestion
void AttributedMaze::remove_invalid_maze_points(std::vector<MazePoint>& tiles) {
```
Please change it to snake_case

File: day_10/task_2/pipe_maze.cpp [hunk 4]
```diff
@@ -73,31 +74,31 @@ std::pair<MazePoint, MazePoint> AttributedMaze::find_next_neighbors(const MazePo
     const MazePoint current_position(maze_point);
     std::vector<MazePoint> neighbors;
 
-    if(is_connected_to_north(
+    if((current_position.first - 1)< maze_.size() && is_connected_to_north(
```
Comment (Rosein) [inline day_10/task_2/pipe_maze.cpp]:
```suggestion
    if(static_cast<std::size_t>(current_position.first - 1) < maze_.size() && is_connected_to_north(
```
1. Add static_cast to make the conversion explicit
2. Add space bar before '<' to make the style consistent.

According to 1. - please watch this: https://www.youtube.com/watch?v=T97QJ0KBaBU to understand the reason
According to 2. - have you applied clang formatter on this PR?

File: day_10/task_2/pipe_maze.cpp [hunk 5]
```diff
@@ -73,31 +74,31 @@ std::pair<MazePoint, MazePoint> AttributedMaze::find_next_neighbors(const MazePo
...
            maze_[current_position.first][current_position.second].tile) &&
        is_connected_to_south(
            maze_[to_north(current_position).first][current_position.second].tile))
     {
         neighbors.push_back(MazePoint(to_north(current_position)));
     }
 
-    if(is_connected_to_south(
+    if((current_position.first +1 )< maze_.size() && is_connected_to_south(
```
Comment (Rosein) [inline day_10/task_2/pipe_maze.cpp]:
```suggestion
    if(static_cast<std::size_t>(current_position.first + 1) < maze_.size() && is_connected_to_south(
```

Comment (Rosein) [reply day_10/task_2/pipe_maze.cpp]: Rationale as previous comment

File: day_10/task_2/pipe_maze.cpp [hunk 6]
```diff
@@ -73,31 +74,31 @@ std::pair<MazePoint, MazePoint> AttributedMaze::find_next_neighbors(const MazePo
...
            maze_[current_position.first][current_position.second].tile) &&
        is_connected_to_north(
            maze_[to_south(current_position).first][current_position.second].tile))
     {
         neighbors.push_back(MazePoint(to_south(current_position)));
-    }
+    }
```
Comment (Rosein) [inline day_10/task_2/pipe_maze.cpp]:
```suggestion
    }
```

Comment (Rosein) [reply day_10/task_2/pipe_maze.cpp]: Remove trailing white spaces

File: day_10/task_2/pipe_maze.cpp [hunk 7]
```diff
@@ -73,31 +74,31 @@ std::pair<MazePoint, MazePoint> AttributedMaze::find_next_neighbors(const MazePo
...
            maze_[current_position.first][current_position.second].tile) &&
        is_connected_to_north(
            maze_[to_south(current_position).first][current_position.second].tile))
     {
         neighbors.push_back(MazePoint(to_south(current_position)));
-    }
+    }   
 
-    if(is_connected_to_east(
+    if(current_position.second + 1 < maze_[0].size() && is_connected_to_east(
```
Comment (Rosein) [inline day_10/task_2/pipe_maze.cpp]:
```suggestion
    if(static_cast<std::size_t>(current_position.second + 1) < maze_[0].size() && is_connected_to_east(
```

Implicit conversion is bad smell

File: day_10/task_2/pipe_maze.cpp [hunk 8]
```diff
@@ -73,31 +74,31 @@ std::pair<MazePoint, MazePoint> AttributedMaze::find_next_neighbors(const MazePo
...
            maze_[current_position.first][current_position.second].tile) &&
        is_connected_to_west(
            maze_[current_position.first][to_east(current_position).second].tile))
     {
         neighbors.push_back(MazePoint(to_east(current_position)));
     }
 
-    if(is_connected_to_west(
+    if(current_position.second -1 < maze_[0].size() && is_connected_to_west(
```
Comment (Rosein) [inline day_10/task_2/pipe_maze.cpp]:
```suggestion
    if(static_cast<std::size_t>(current_position.second - 1) < maze_[0].size() && is_connected_to_west(
```

File: day_10/task_2/pipe_maze.cpp [hunk 9]
```diff
@@ -211,16 +210,22 @@ void AttributedMaze::set_tile_at(const MazePoint& point, const Tile& tile)
 
 bool AttributedMaze::is_in_maze(const MazePoint& point) const
 {
-    return point.first >= 0 and point.first < static_cast<int>(maze_[0].size()) and
-        point.second >= 0 and point.second < static_cast<int>(maze_.size());
+    return point.first >= 0 and point.first < static_cast<int>(maze_.size()) and point.second >= 0 and point.second < static_cast<int>(maze_[0].size());
```
Comment (Rosein) [inline day_10/task_2/pipe_maze.cpp:213]: [optional] 

This change should be commited separately with title in commit like this:

`FIX(AttributedMaze): Swap the dimensions in is_in_maze`

File: day_10/task_2/pipe_maze.cpp (hunk 3 above)
Comment (Rosein) [reply day_10/task_2/pipe_maze.cpp]: [optional]

This function could be also delivered as separate commit with FIX prefix with description of the reason

File: day_10/task_2/pipe_maze.cpp [hunk 10]
```diff
@@ -270,3 +276,86 @@ void AttributedMaze::color_neighbor(const MazePoint& loop_point, const MazePoint
...
+            }
+            if(single_tile.state == State::Loop)
+            {
+                std::cout << "X";
+            }
+
+        }
+        std::cout << std::endl;
+    }
+}
```
Comment (Rosein) [inline day_10/task_2/pipe_maze.cpp:306]: [ optional ]

Deliver this function as separated commit, which is not neccesary part of the solution. So maybe DEBUG prefix could be fit. Of course such commit would need to include `#include <iostream` directive

File: day_10/task_2/pipe_maze.cpp [hunk 11]
```diff
@@ -270,3 +276,86 @@ void AttributedMaze::color_neighbor(const MazePoint& loop_point, const MazePoint
...
+                return State::RightColor;
+            }
+            if(maze_[i][j].state == State::RightColor)
+            {
+                return State::LeftColor;
+            }
+        }
+    }
+    return State::Undefined;
+}
```
Comment (Rosein) [inline day_10/task_2/pipe_maze.cpp:343]:
```suggestion
State AttributedMaze::find_which_color_is_closed()
{
    for(auto i = 0U; i < maze_.size(); i++)
    {
        for(auto j = 0U; j < maze_[i].size(); j++)
        {
            if(maze_[i][j].state == State::LeftColor)
            {
                return State::RightColor;
            }
            if(maze_[i][j].state == State::RightColor)
            {
                return State::LeftColor;
            }
        }
    }
    return State::Undefined;
}
```

[optional]

Please consider the version of algorithm which would cover all possible input cases

File: day_10/task_2/pipe_maze.cpp [hunk 12]
```diff
@@ -270,3 +276,86 @@ void AttributedMaze::color_neighbor(const MazePoint& loop_point, const MazePoint
...
+    State closed_color = find_which_color_is_closed();
+    int counter{0};
+    for(auto i = 0U; i < maze_.size(); i++)
+    {
+        for(auto j = 0U; j < maze_[i].size(); j++)
+        {
+            if(maze_[i][j].state == closed_color)
+            {
+                ++counter;
+            }
```
Comment (Rosein) [inline day_10/task_2/pipe_maze.cpp]: Please refactor this with [count_if](https://en.cppreference.com/w/cpp/algorithm/count)

File: day_10/task_2/pipe_maze.hpp [hunk 13]
```diff
@@ -35,8 +35,17 @@ struct AttributedMaze
...
+    void removeInvalidMazePoints(std::vector<MazePoint>& tiles);
+
+
+    void print_states();
 
     std::vector<std::vector<AttributedMazePoint>> maze_{};
+
 };
 
 AttributedMaze transform_to_attributed_maze(const PipeMaze& pipe_maze);
```
Comment (Rosein) [inline day_10/task_2/pipe_maze.hpp]: Missing newline at the end

File: debug_features/debug_features.hpp [hunk 14]
```diff
@@ -2,7 +2,7 @@
 
 #include <iomanip>
 #include <iostream>
-const bool kIsDebugOn = false;
```
Comment (Rosein) [inline debug_features/debug_features.hpp]: Should not be part of the PR

File: day_10/task_2/tests/pipe_maze_tests.cpp [hunk 15]
```diff
@@ -221,14 +221,23 @@ TEST(AttributedMazeTest, GivenAttributedMazeWithMarkedLoop_WhenCalculateEnclosed
...
+
+//  PipeMaze pipe_maze = {
+//         "L--JL--J",
+//         "|..||..|",
+//         "|L-7F-J|",
+//         "||....||",
+//         "||....||",
+//         "|F----7|",
+//         "S------7",
+//     };
```
Comment (Rosein) [inline day_10/task_2/tests/pipe_maze_tests.cpp]: Please remove it or transform it into the unit test.

File: day_10/task_2/pipe_maze.cpp (hunk 3 above)
Comment (Rosein) [reply day_10/task_2/pipe_maze.cpp]: Really good catch that you spot the need to add this function and you implemented it! ❤️

**Review by Rosein — CHANGES_REQUESTED** [review]
Good job! Please consider minor suggestions

File: day_10/task_2/BUILD (hunk 1 above)
Comment (dariannax3) [reply day_10/task_2/BUILD]: done

File: day_10/task_2/BUILD (hunk 2 above)
Comment (dariannax3) [reply day_10/task_2/BUILD]: done

File: day_10/task_2/pipe_maze.cpp (hunk 3 above)
Comment (dariannax3) [reply day_10/task_2/pipe_maze.cpp]: thank you <3

File: day_10/task_2/pipe_maze.cpp (hunk 4 above)
Comment (dariannax3) [reply day_10/task_2/pipe_maze.cpp]: done

File: day_10/task_2/pipe_maze.cpp (hunk 5 above)
Comment (dariannax3) [reply day_10/task_2/pipe_maze.cpp]: done

File: day_10/task_2/pipe_maze.cpp (hunk 6 above)
Comment (dariannax3) [reply day_10/task_2/pipe_maze.cpp]: done

File: day_10/task_2/pipe_maze.cpp (hunk 7 above)
Comment (dariannax3) [reply day_10/task_2/pipe_maze.cpp]: done

File: day_10/task_2/pipe_maze.cpp (hunk 9 above)
Comment (dariannax3) [reply day_10/task_2/pipe_maze.cpp:213]: let it leave in one commit

File: day_10/task_2/pipe_maze.cpp (hunk 12 above)
Comment (dariannax3) [reply day_10/task_2/pipe_maze.cpp]: done

File: day_10/task_2/pipe_maze.hpp (hunk 13 above)
Comment (dariannax3) [reply day_10/task_2/pipe_maze.hpp]: done

File: debug_features/debug_features.hpp (hunk 14 above)
Comment (dariannax3) [reply debug_features/debug_features.hpp]: done

File: day_10/task_2/tests/pipe_maze_tests.cpp (hunk 15 above)
Comment (dariannax3) [reply day_10/task_2/tests/pipe_maze_tests.cpp]: new test added

**Review by Rosein — APPROVED** [review]
